### PR TITLE
chore: Bump to Python3.10

### PR DESCRIPTION
**Pull request by EugeneTorap**
### SUMMARY
Running CI integration and unit tests for Py3.10 and Py3.11.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: setup.py
Comment (villebro) [inline setup.py:193]: I would recommend leaving explicit support for 3.9 for a while longer, as we've only just deprecated 3.8.

Comment (EugeneTorap) [reply setup.py:193]: If we want to support Py3.9 then we don't need this PR

**Comment by john-bodley** [comment]
@betodealmeida, @eschutho, @michael-s-molina, @villebro et al. do we have a policy in terms of what versions of Python we support (I wasn’t able to find anything in the repo via a quick code search)? If not we probably should formalize this, i.e., 

- The [official supported versions](https://devguide.python.org/versions/#supported-versions) (excluding `main`)
- The _n_ latest versions

**Comment by betodealmeida** [comment]
> @betodealmeida, @eschutho, @michael-s-molina, @villebro et al. do we have a policy in terms of what versions of Python we support (I wasn’t able to find anything in the repo via a quick code search)? If not we probably should formalize this, i.e.,
> 
>     * The [official supported versions](https://devguide.python.org/versions/#supported-versions) (excluding `main`)
> 
>     * The non-prerelease official supported versions
> 
>     * The _n_ latest versions
> 
>     * …

Agree, @john-bodley, we should have this formalized. My suggestions would be every major version of Superset should support the latest 3 major versions of Python available at the time (so 3.9-3.11 today), and probably when 4.0 is out we'll support 3.10-3.12.

**Comment by michael-s-molina** [comment]
> My suggestions would be every major version of Superset should support the latest 3 major versions of Python available at the time (so 3.9-3.11 today), and probably when 4.0 is out we'll support 3.10-3.12.

This is a great topic for the next Town Hall @rusackas

**Comment by mdeshmu** [comment]
@michael-s-molina Is there any decision made on officially supported Python versions?

**Comment by EugeneTorap** [comment]
Hi @betodealmeida @john-bodley @michael-s-molina @mdeshmu We have a problem with `Py3.11` and `prophet==1.1.1` but we can't update the prophet because that [nasty bug](https://github.com/facebook/prophet/issues/2354) hasn't been fixed yet.

**Comment by mdeshmu** [comment]
@EugeneTorap I will follow up on [the bug](https://github.com/facebook/prophet/issues/2354). 

Someone said [here](https://github.com/apache/superset/pull/25017#issuecomment-1706293216) that combination of prophet 1.1.4 and holidays 0.32 is working. Can you check that? 

Also, there is another [potential Redis issue](https://github.com/apache/superset/issues/25471#issuecomment-1742337651) with Python 3.11.

**Comment by mdeshmu** [comment]
@EugeneTorap prophet 1.1.5 is released and someone said here it works for them: https://github.com/facebook/prophet/issues/2354#issuecomment-1778888086

**Comment by EugeneTorap** [comment]
Thanks @mdeshmu for this information! Can you create a small PR to update the prophet to version 1.1.5?

**Comment by villebro** [comment]
@mdeshmu @EugeneTorap I believe I tested the new prophet version, but it was still causing issues for me. But let me double check, and raise a bump PR if it does in fact work.

**Comment by rusackas** [comment]
Nice to see you again @EugeneTorap :) Thanks for revisiting this!

**Review by craig-rueda — APPROVED** [review]
LGTM

**Comment by mistercrunch** [comment]
I think you'll need to rebase because the required checks have change since you originally branched off. 

My 2 cents on python is Superset should be able to dictate a single version we fully support. It's not like it's a library that you may want to use as part of an app, it's its own app and therefore can be firm on what it runs on. Running against next/latest version in CI may be reasonable too.

File: .github/workflows/docker-release.yml
Comment (mistercrunch) [inline .github/workflows/docker-release.yml:47]: is removing this exclude clause deliberate? Actually might be reasonable to remove this from releases, but not from PRs (below)

Comment (mistercrunch) [inline .github/workflows/docker-release.yml:47]: is removing this exclude clause deliberate? Actually might be reasonable to remove this from releases, but not from PRs (below)

File: .github/workflows/docker.yml
Comment (mistercrunch) [inline .github/workflows/docker.yml:22]: I'd say we keep the exclude here, this runs for every commit in every PR, and on every merge onto master, and arm is quite slow/expensive because it's emulated with QEMU

File: Dockerfile
Comment (mistercrunch) [inline Dockerfile:21]: Should we pin to the micro version here or just point to the minor like we did before? There are pros/cons here, but we may miss out on security updates, and we don't have a process like dependabot to monitor and bump this. I'd say we stick to the previous strategy to point to the minor release image.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Hey, sorry to override, but have a few nuggets I'd like tackled here. Looked like you have to rebase too. Thanks for doing this!

File: Dockerfile
Comment (mistercrunch) [reply Dockerfile:21]: To clarify, I'd vote to point to `3.10-slim-bookworm` here. It looks available here https://hub.docker.com/_/python

**Comment by rusackas** [comment]
@EugeneTorap wondering if you'll indeed have time to tackle the above comments and rebase this PR. Also, send us a message by email or Slack... we'd tried to reach you by email on another matter, but haven't heard back. Thanks!

**Comment by mistercrunch** [comment]
I'd say let's merge this here first and rebase on top of it as it brings setting up the backend in a reusable action, and sets a default -> https://github.com/apache/superset/pull/27505

**Comment by mistercrunch** [comment]
@EugeneTorap I added a commit to try and help moving past the python dependencies issues.

**Comment by mistercrunch** [comment]
About the required checks that are not triggering, I can provide some context, first you'll want to bump the requirements here -> https://github.com/apache/superset/blob/master/.asf.yaml#L70-L78, these define the required checks to merge onto `master`.

Now since GitHub looks at the master branch to define what's needed (instead of looking at your branch/PR), you want to add a bunch of no-op checks as part of your PR so that the check will be green. You can do that here: https://github.com/apache/superset/blob/master/.github/workflows/no-op.yml and we can clean those up later once this is merge.

**Review by mistercrunch — APPROVED** [review]

**Comment by mistercrunch** [comment]
for the docker step you'll probably need this -> https://github.com/apache/superset/pull/27615, feel free to wait to merge + rebase or cherry-pick into this PR

**Comment by mistercrunch** [comment]
Nice! Oh one more thing, can you add an entry on `./UPDATING.md` about Python 3.10 being the new minimum python version to run Superset and referencing this PR.

Otherwise LGTM, but for something that important we could use more approvals.

**Comment by EugeneTorap** [comment]
@john-bodley @villebro @dpgaspar @rusackas can you also approve & merge this PR?

**Review by villebro — APPROVED** [review]
Nice, LGTM!

File: setup.py
Comment (villebro) [inline setup.py:211]: I missed this just now, can we make it 
```suggestion
    python_requires="~=3.10",
```

**Review by michael-s-molina — CHANGES_REQUESTED** [review]
This is breaking change for folks using Python 3.9. It should become a proposal for the next breaking window.

**Comment by villebro** [comment]
> This is breaking change for folks using Python 3.9. It should become a proposal for the next breaking window.

@michael-s-molina thanks for reviewing, I suspected this would be the case.. In that case, I propose the following:
- We change this PR to continue support for 3.9
- Open a new card on the 5.0 board that removes 3.9 when the window opens again, assigning it to @EugeneTorap 

Thoughts?

**Comment by michael-s-molina** [comment]
> We change this PR to continue support for 3.9

That would work.

> Open a new card on the 5.0 board that removes 3.9 when the window opens again, assigning it to @EugeneTorap
Thoughts?

Added [here](https://github.com/orgs/apache/projects/292?pane=issue&itemId=57390800).

**Comment by mistercrunch** [comment]
Ok, so main / published docker builds start pointing to`3.10`, and docker-compose / helm chart start pointing to 3.10 as a result,... But we still:
- officially support 3.9 in setup.py
- run all or some of the tests on both `3.9` and `3.10`?

I noticed we have a lot of "matrix" in GHA that are 3.9 only, would we want to have some or all of those do both?

File: setup.py
Comment (mistercrunch) [inline setup.py:217]: ok so we want to keep this line for the time being

File: .github/actions/setup-backend/action.yml
Comment (mistercrunch) [inline .github/actions/setup-backend/action.yml:27]: redundant with the next step, probably a bad merge conflict (?)

File: scripts/build_docker.py
Comment (mistercrunch) [inline scripts/build_docker.py:152]: optional, but we could add a `py39` preset here while we support 3.9 (until we remove it in setup.py)

**Review by mistercrunch — CHANGES_REQUESTED** [review]

File: setup.py
Comment (EugeneTorap) [reply setup.py:217]: @michael-s-molina Do we want to wait until version 4.0 is released? And after this, will we merge this PR?

Comment (michael-s-molina) [reply setup.py:217]: We don't have any dependency with 4.0 to merge this PR. If you address both comments from @mistercrunch, we'll be able to merge it.

File: setup.py
Comment (michael-s-molina) [inline setup.py:211]: This should also be reverted given that:

> The `python_requires` field is used in the setup.py file of a Python package to specify the minimum version(s) of Python required to install and use the package.

and 3.9 would be the minimum version.

File: setup.py
Comment (EugeneTorap) [reply setup.py:217]: @mistercrunch Can you push the changes and merge it? Because I don't have time for it now

**Review by mistercrunch — APPROVED** [review]
I made the suggested changes. Surprising how editing on-the-web is pretty viable now with GitHub. 

@michael-s-molina I'll need an approval from you to merge since you're still in "Request changes" from a previous review

**Review by michael-s-molina — APPROVED** [review]

**Comment by mistercrunch** [comment]
UPDATE YOUR VENVs EVERYONE! :) 

giphy

Or just use docker-compose for your dev env

**Pull request merged**
